docs/src/reference/the-traversal.asciidoc

```diff
@@ -4099,8 +4100,12 @@ The provided predicates are outlined in the table below and are used in various
 | `TextP.notStartingWith(string)` | Does the incoming `String` not start with the provided `String`?
 | `TextP.notEndingWith(string)` | Does the incoming `String` not end with the provided `String`?
 | `TextP.notContaining(string)` | Does the incoming `String` not contain the provided `String`?
+| `TextP.regex(string)` | Does the incoming `String` match the regular expression in the provided `String`?
```

Comment:
I have a question on the name of the predicate. Please feel free to point me to the email thread is a discussion has already taken place.

While adding new syntax to Gremlin, we should try to maintain consistency with other query languages / programming languages. This will help reduce the learning curve for users of Gremlin.

Keeping the above tenet in mind for adding new syntax, I would favour a more widely used syntax for regex match such as:

TextP.like() -> used by SQL
or
TextP.match() -> used by Java / Python

I would favor the former (matches()/match()).

Also, we don't need a special negation syntax, since the user can do `P.not(TextP.regex())`

Reply:
i'd have preferred `match()` as well but it conflicts with `match()`-step. 

> Also, we don't need a special negation syntax, since the user can do

The negate is there to satisfy the `Predicate.negate()` with a non-lambda for serialization purposes. we might add a comment thought to say that it typically isn't recommended to use it.

`like()` is actually a nice alternative as it makes the `notLike()` easier. `like()` isn't self-explaining like `regex()` but perhaps that's not a problem. I'd be happy with `like()`. I suppose i'd also be in favor of `matches()` which is amenable to the negation.

Reply:
I took "inspiration" from other Graphs and query languages. JanusGraph and Titan use the "regex" phrase as does SPARQL with its  `FILTER REGEX`. Cypher uses the Ruby style of `=~`  so neither a `like` or a `regex`. I considered `match` but did not go with it for the reasons discussed. We could go with `like` but `like` is not really a regex matcher in SQL it's a more basic syntax. There have been e-mails on this that date back to when the original TextP.startingWith work was done but I do not have links handy. There is also a Jira. I would still prefer to go with `regex` as that makes it very clear what the step is going to do to my mind.

Reply:
My main concern with using `TextP.regex()` syntax is that it doesn't disambiguate between the following two definitions:
1. Returns true if the entire input matches the regex pattern.
2. Returns true if part of the string matches the regex pattern.

Consider examples:

If we use definition#1
```
Input string "abc"
RegEx "a"
Result: false
```

If we use definition#2
```
Input string "abc"
RegEx "a"
Result: true
```

SQL operator LIKE follows definition#1 above. 

OpenCypher uses Java's regular expression [3] which follows definition#1 above.

Sparql regex uses Xpath's fn:matches [1] which follows definition#2 above [2]

In our case, we are using definition#1 similar to OC `~=` and SQL's `LIKE` operator whereas using `TextP.regex()` syntax indicates similarity to SPARQL which points to definition#2. That is the problem in using `TextP.regex()` as the syntax.

Moving ahead, we could either:
1. Continue using `TextP.regex()` syntax but clarify the behaviour explicitly that we are using definition#1 above and should not be confused with definition#2 used by other languages like SPARQL.
2. Use `TextP.like()` syntax which is more consistent with user experience.

I would favor option#2 for consistency with other languages and got providing gotchas for users of TinkerPop.

[1] SPARQL uses XPath's fn:matches -> https://www.w3.org/TR/sparql11-query/#func-regex
[2] *Unless the metacharacters ^ and $ are used as anchors, the string is considered to match the pattern if any substring matches the pattern.* Quoted from https://www.w3.org/TR/xpath-functions/#func-matches
[3] https://neo4j.com/docs/cypher-manual/current/clauses/where/#query-where-regex